### PR TITLE
Outlaw recursive types: occurs check

### LINKED ISSUE #394: Recursive types are not being outlawed by typechecker
This program should not typecheck, because `gonzo` has type `mu a. Sum Bool a`.  
```
feature salary
~> fold gonzo = Left True : Right gonzo
~> 0
```

Test failure details:
```
=== prop_convert_is_well_typed from test/Icicle/Test/Source/Convert.hs:30 ===
*** Failed! Falsifiable (after 351 tests):
QueryWithFeature {qwfQuery = Query {contexts = [LetFold () (Fold {foldBind = Name {nameHash = -3536051112772635155, nameBase = NameBase (Variable "gonzo")}, foldInit = Nested () (Query {contexts = [Let () (Name {nameHash = 1731273602156757220, nameBase = NameBase (Variable "chef")}) (App () (Prim () (PrimCon ConLeft)) (Prim () (PrimCon ConTrue)))], final = Var () (Name {nameHash = 1731273602156757220, nameBase = NameBase (Variable "chef")})}), foldWork = App () (Prim () (PrimCon ConRight)) (Var () (Name {nameHash = -3536051112772635155, nameBase = NameBase (Variable "gonzo")})), foldType = FoldTypeFoldl})], final = Prim () (Lit (LitDouble 3.8927332184613013))}, qwfNow = Just (Name {nameHash = -2746986193221104132, nameBase = NameBase (Variable "animalnow")}), qwfOutput = OutputName "animal", qwfFeatureT = UnitT, qwfFeatureN = Name {nameHash = -4912334399949111077, nameBase = NameBase (Variable "animalfeaturename")}, qwfTimeName = Name {nameHash = -4180041470634012872, nameBase = NameBase (Variable "animaltimename")}}

feature animalfeaturename
~> fold gonzo = (let chef = Left True
                 ~> chef) : Right gonzo
~> 3.8927332184613013

Program (core$0 : Unit, core$1 : FactIdentifier, core$2 : Time, core$3 : SNAPSHOT_TIME)
Precomputations:
  core$6               = 3.8927332184613013
Streams:
  STREAM_FOLD (gonzo$core$5 : (Sum Bool (Sum Bool (Sum Bool Unit))))
    INIT:
        let chef$core$4 = left# (True)
         in chef$core$4
    KONS:
      right# gonzo$core$5
Postcomputations:
Returning:
  animal               = core$6

Left (ProgramErrorStream (StreamErrorExp (ExpErrorApp (XPrim () (PrimMinimal (PrimConst (PrimConstRight BoolT (SumT BoolT (SumT BoolT UnitT)))))) (XVar () (Name {nameHash = -1034977650126524493, nameBase = NameMod (Variable "gonzo") (NameMod (Variable "core") (NameBase (Variable "5")))})) (FunT [FunT [] (SumT BoolT (SumT BoolT UnitT))] (SumT BoolT (SumT BoolT (SumT BoolT UnitT)))) (FunT [] (SumT BoolT (SumT BoolT (SumT BoolT UnitT)))))))
```

### PR DESCRIPTION
**Pull request by amosr**
fix #394. surprisingly simple.

### CODE REVIEW

File: src/Icicle/Source/Type/Subst.hs [hunk 1]
```diff
@@ -149,10 +150,21 @@ unifyT t1 t2
      | TypeVar b <- t2
      , a == b
      -> return $ Map.empty
+
     TypeVar a
+     -- Occurs check.
+     -- TODO: it would be nice to have a better error message than just
+     -- "Could not unify".
+     -- Something specifically about recursive types would be ideal.
```
Comment (jystic) [inline src/Icicle/Source/Type/Subst.hs]: We should make a ticket for this

**Comment by jystic** [comment]
🔞

**Comment by tranma** [comment]
since these type errors are in Source, could we also include the names in the error message?  e.g. `foo$simp$2 has type Int, but is required to be Double`

**Comment by amosr** [comment]
yeah, jacob and I were just talking about this. I'd like a datatype which is like a path, something like
```
data PathElement = Binding Name | FilterPred | GroupKey
```
or something, and then your constraint has a list of these `PathElement`, so you can say
```
In the binding foo$simp, in the filter predicate, the type blah is not blah
```

**Pull request merged**